[EDGE-1520] Add pipeline templates rest & graphql api documentation (#2479)
buildkite/docs: pages/apis/rest_api/pipeline_templates.md

Reviewer: I think it would be better if `buildkite.com/organizations/buildkite/` were replaced with `buildkite.com/organizations/acme-inc/` or similar like our other REST API docs seem to use.

Author: Updated in [6e8fe10](https://github.com/buildkite/docs/pull/2479/commits/6e8fe10b3f7d6454e5763f8ba26add103c957706).
Doesn't seem like there's a set convention but I've changed it to `acme-inc` to remain somewhat consistent with the other pages :)

> # Pipeline templates API

Reviewer: @jonathanly my expectation when loading up the docs was that this would be under "Pipelines REST API > Templates" or "Pipelines REST API > Pipeline Templates".

Is there a particular reason why this warrants a top-level section?

Author: On looking at it again, it doesn't make sense at all so I've moved the page under Pipelines REST API ([e7d11cb](https://github.com/buildkite/docs/pull/2479/commits/e7d11cb314de0cdb038a225b783d28bc4dc7e420))

> 📘 Enterprise feature
> Pipeline templates are only available on an [Enterprise](https://buildkite.com/pricing) plan.

The pipeline templates API allows admins to create and manage pipeline templates for an organization.
Non-admins will only be permitted to read pipeline templates marked as available by organization admins.

> 

Reviewer: I think this should link to the pipeline templates documentation so that a user can learn what they are actually used for.

Author: I like that suggestion, updated this in [39d7dd7](https://github.com/buildkite/docs/pull/2479/commits/39d7dd730b1b5c0910b26e9120bd3c3432ebbb7b)

> Non-admins will only be permitted to read pipeline templates marked as available by organization admins.

Reviewer 2: I think we should use code formatting for "available:"

```suggestion
Non-admins will only be permitted to read pipeline templates marked as `available` by organization admins.
```

Is "read" the right description for what `available` allows for non-admins? In the field description it mentions that when `available` is true, non-admins can assign templates to pipelines.

Author: In the context of the pipeline templates rest api, non admin users will only be able to perform `#index` and `#show` read actions on templates marked as `available`.
They'll be able to assign available templates to pipelines via the pipelines rest api :)

Reviewer 2: Okay, so the key difference is not being able to create, update, or delete the actual templates. How about:

> Non-admins can only read or assign pipeline templates marked as `available` by organization admins.

## Pipeline template data model

<table class="responsive-table">
<tbody>
  <tr><th><code>uuid</code></th><td>UUID of the pipeline template</td></tr>
  <tr><th><code>graphql_id</code></th><td><a href="/docs/apis/graphql-api#graphql-ids">GraphQL ID of the pipeline template</a></td></tr>
  <tr><th><code>name</code></th><td>Name of the pipeline template</td></tr>
  <tr><th><code>description</code></th><td>Description of the pipeline template</td></tr>
  <tr><th><code>configuration</code></th><td>YAML step configuration for the pipeline template</td></tr>

>   <tr><th><code>available</code></th><td>When set to true, the pipeline template can assigned to pipelines by non-admins<br><em>Default:</em> <code>false</code></td></tr>

Reviewer 2: Missing code formatting:

```suggestion
  <tr><th><code>available</code></th><td>When set to <code>true</code>, the pipeline template can assigned to pipelines by non-admins<br><em>Default:</em> <code>false</code></td></tr>
```

>   <tr><th><code>available</code></th><td>When set to <code>true</code>, the pipeline template can assigned to pipelines by non-admins<br><em>Default:</em> <code>false</code></td></tr>

Reviewer 2: One last instance:

```suggestion
  <tr><th><code>available</code></th><td>When set to <code>true</code>, non-admins can assign the pipeline template to pipelines<br><em>Default:</em> <code>false</code></td></tr>
```

Author: Did a full text search on the code just to be sure.. 😆
That ought to be it!

  <tr><th><code>url</code></th><td>Canonical API URL of the pipeline template</td></tr>
  <tr><th><code>web_url</code></th><td>URL of the pipeline template on Buildkite</td></tr>
  <tr><th><code>created_at</code></th><td>When the pipeline template was created</td></tr>

>   <tr><th><code>created_by</code></th><td>User who created the pipeline template</td></tr>

Reviewer 2: Is there a data structure we can link to for this object? Maybe:

```suggestion
  <tr><th><code>created_by</code></th><td><a href="/docs/apis/rest-api/user">User</a> who created the pipeline template</td></tr>
```

  <tr><th><code>updated_at</code></th><td>When the pipeline template was created</td></tr>

>   <tr><th><code>updated_by</code></th><td>User who last updated the pipeline template</td></tr>

Reviewer 2: Similarly, should we link to the same User object?

</tbody>
</table>

## List pipeline templates

Returns a list of an organization's pipeline templates.

```bash
curl "https://api.buildkite.com/v2/organizations/{org.slug}/pipeline-templates"
```

```json
[
  {
    "uuid": "018a86cc-db73-7d15-8c68-5023cf8d64c3",
    "graphql_id": "UGlwZWxpbmVUZW1wbGF0ZS0tLTAxOGE4NmNjLWRiNzMtN2QxNS04YzY4LTUwMjNjZjhkNjRjMw==",
    "name": "Build template",
    "description": "Shared build steps configuration",
    "configuration": "steps:\n  - label: \":hammer: Build\"\n    command: \"scripts/build.sh\"",
    "available": false,
    "url": "http://api.buildkite.localhost/v2/organizations/buildkite/pipeline-templates/018a86cc-db73-7d15-8c68-5023cf8d64c3",

>     "web_url": "http://buildkite.localhost/organizations/buildkite/pipeline-templates/018a86cc-db73-7d15-8c68-5023cf8d64c3",

Reviewer: I note that [our examples for things like Clusters](https://buildkite.com/docs/apis/rest-api/clusters) use production URLs rather than local ones, we should do the same here.

Author: Edited the urls in this commit ([e7d11cb](https://github.com/buildkite/docs/pull/2479/commits/e7d11cb314de0cdb038a225b783d28bc4dc7e420))

  },
   "updated_at": "2023-06-12T04:17:55.867Z",
  "updated_by": {
    "id": "3d3c3bf0-7d58-4afe-8fe7-b3017d5504de",
    "graphql_id": "VXNlci0tLTNkM2MzYmYwLTdkNTgtNGFmZS04ZmU3LWIzMDE3ZDU1MDRkZQo=",
    "name": "Sam Kim",
    "email": "sam@example.com",
    "avatar_url": "https://www.gravatar.com/avatar/example",
    "created_at": "2013-05-03T04:17:55.867Z"
  }
}
```

Required scope: `read_pipeline_templates`

Success response: `200 OK`

## Create a pipeline template

```bash
curl -X POST "https://api.buildkite.com/v2/organizations/{org.slug}/pipeline-templates" \
  -H "Content-Type: application/json" \
  -d '{
    "name": ":hammer: Build",
    "description": "Shared build steps configuration",
    "configuration": "steps:\n  - label: \":hammer: Build\"\n    command: \"scripts/build.sh\"",
    "available": true,
  }'
```

```json
{
  "uuid": "018a86cc-db73-7d15-8c68-5023cf8d64c3",
  "graphql_id": "UGlwZWxpbmVUZW1wbGF0ZS0tLTAxOGE4NmNjLWRiNzMtN2QxNS04YzY4LTUwMjNjZjhkNjRjMw==",
  "name": "Build template",
  "description": "Shared build steps configuration",
  "configuration": "steps:\n  - label: \":hammer: Build\"\n    command: \"scripts/build.sh\"",

>   "available": false,

Reviewer 2: This was set to `true` in the payload to create the template, should it also be true here?

```suggestion
  "available": true,
```

  },
  "updated_at": "2023-06-12T04:17:55.867Z",
  "updated_by": {
    "id": "3d3c3bf0-7d58-4afe-8fe7-b3017d5504de",
    "graphql_id": "VXNlci0tLTNkM2MzYmYwLTdkNTgtNGFmZS04ZmU3LWIzMDE3ZDU1MDRkZQo=",
    "name": "Sam Kim",
    "email": "sam@example.com",
    "avatar_url": "https://www.gravatar.com/avatar/example",
    "created_at": "2013-05-03T04:17:55.867Z"
  }
}
```

Required [request body properties](/docs/api#request-body-properties):

<table class="responsive-table">
<tbody>
  <tr>
    <th><code>name</code></th>
    <td>Name for the pipeline template.<br><em>Example:</em> <code>"Build template"</code></td>
  </tr>
  <tr>
    <th><code>configuration</code></th>
    <td>YAML step configuration for the pipeline template.<br><em>Example:</em> <code>"steps:\n  - command: "scripts/build.sh"</code></td>
  </tr>
</tbody>
</table>

Optional [request body properties](/docs/api#request-body-properties):

<table class="responsive-table">
<tbody>
  <tr>
    <th><code>description</code></th>
    <td>Description for the pipeline template.<br><em>Example:</em> <code>"Shared build steps configuration"</code></td>
  </tr>
  <tr>
    <th><code>available</code></th>

>     <td>When set to true, the pipeline template can assigned to pipelines by non-admins.<br><em>Example:</em> <code>false</code></td>

Reviewer 2: Missing code formatting + active voice:

```suggestion
    <td>When set to <code>true</code>, non-admins can assign the pipeline template to pipelines.<br><em>Example:</em> <code>false</code></td>
```

>     <td>When set to <code>true</code>, non-admins can assign the pipeline template can assigned to pipelines.<br><em>Example:</em> <code>false</code></td>

Reviewer 2: Looks like this change didn't get applied quite right. Should be:

```suggestion
    <td>When set to <code>true</code>, non-admins can assign the pipeline template to pipelines.<br><em>Example:</em> <code>false</code></td>
```

I think this mistake is made in a few places.

Author: Hopefully that last commit I just pushed up covered the rest of it 🤞

  </tr>
</tbody>
</table>

Required scope: `write_pipeline_templates`

Success response: `201 Created`

Error responses:

<table class="responsive-table">
<tbody>
  <tr><th><code>422 Unprocessable Entity</code></th><td><code>{ "message": "Validation failed: Reason for failure" }</code></td></tr>
</tbody>
</table>

## Update a pipeline template

```bash
curl -X PATCH "https://api.buildkite.com/v2/organizations/{org.slug}/pipeline-templates/{uuid}" \
  -H "Content-Type: application/json" \
  -d '{ "available": true }'
```

```json
{
  "uuid": "018a86cc-db73-7d15-8c68-5023cf8d64c3",
  "graphql_id": "UGlwZWxpbmVUZW1wbGF0ZS0tLTAxOGE4NmNjLWRiNzMtN2QxNS04YzY4LTUwMjNjZjhkNjRjMw==",
  "name": "Build template",
  "description": "Shared build steps configuration",
  "configuration": "steps:\n  - label: \":hammer: Build\"\n    command: \"scripts/build.sh\"",

>   "available": false,

Reviewer 2: It looks like the update here is setting `available` to `true`:

```suggestion
  "available": true,
```

    "name": "Sam Kim",
    "email": "sam@example.com",
    "avatar_url": "https://www.gravatar.com/avatar/example",
    "created_at": "2023-08-29T10:10:03.000Z"
  },
  "updated_at": "2023-06-12T04:17:55.867Z",
  "updated_by": {
    "id": "3d3c3bf0-7d58-4afe-8fe7-b3017d5504de",
    "graphql_id": "VXNlci0tLTNkM2MzYmYwLTdkNTgtNGFmZS04ZmU3LWIzMDE3ZDU1MDRkZQo=",
    "name": "Sam Kim",
    "email": "sam@example.com",
    "avatar_url": "https://www.gravatar.com/avatar/example",
    "created_at": "2013-05-03T04:17:55.867Z"
  }
}
```

Optional [request body properties](/docs/api#request-body-properties):

<table class="responsive-table">
<tbody>
  <tr>
    <th><code>name</code></th>
    <td>Name for the pipeline template.<br><em>Example:</em> <code>"Build template"</code></td>
  </tr>
  <tr>
    <th><code>description</code></th>
    <td>Description for the pipeline template.<br><em>Example:</em> <code>"Shared build steps configuration"</code></td>
  </tr>
  <tr>
    <th><code>configuration</code></th>
    <td>YAML step configuration for the pipeline template.<br><em>Example:</em> <code>"steps:\n  - command: "scripts/build.sh"</code></td>
  </tr>
  <tr>
    <th><code>available</code></th>

>     <td>When set to true, the pipeline template can assigned to pipelines by non-admins.<br><em>Example:</em> <code>false</code></td>

Reviewer 2: ```suggestion
    <td>When set to <code>true</code>, non-admins can assign the pipeline template to pipelines.<br><em>Example:</em> <code>false</code></td>
```

  </tr>
</tbody>
</table>

Required scope: `write_pipeline_templates`

Success response: `200 OK`

Error responses:

<table class="responsive-table">
<tbody>
  <tr><th><code>422 Unprocessable Entity</code></th><td><code>{ "message": "Validation failed: Reason for failure" }</code></td></tr>
</tbody>
</table>

## Delete a pipeline template

>📘

> > A pipeline template can only be deleted when it is not assigned to any pipelines.

Reviewer 2: I think it's worth reinforcing the action required with a second sentence:

```suggestion
> A pipeline template can only be deleted when it is not assigned to any pipelines. Ensure you remove the pipeline template from all pipelines before trying to delete it.
```
